### DISCUSSION TITLE
Buffer reordering

### DISCUSSION THREAD

**Discussion by Omnikar** [helix-editor/helix#10278] [Ideas]
With the new bufferline feature introduced by #2759, the order of buffers becomes much more apparent and important. As a result, I feel it necessary to have commands that can be used to reorder buffers. Unfortunately, it would appear that the data structure currently used to store buffers, `BTreeMap`, does not allow its data to be in an arbitrary order.

**Comment by emilyyyylime** [comment]
I'm working on implementing this, do you reckon having just a "move currently active document left/right" command would be good? And what keybinds should it use? I was thinking we could change "Select symbol references" to something else and use `<space>h` and `<space>l`

**Comment by emilyyyylime** [comment]
Looking into it Kakoune went with a specialised command that would take a list of buffers and put them at the front in order: https://discuss.kakoune.com/t/brainstorm-for-a-more-flexible-buffer-list/767/9. I think this is a good idea, it should work quite well with command completions. I do feel like there should also be an option to move buffers to the end, should that just be a separate command? like `:reorder-buffers` and `:reorder-buffers-reverse`?

**Comment by emilyyyylime** [comment]
Also in terms of implementation do you think it might be a good idea to use a different collection for this? Because currently my code for swapping 2 documents removes them both from the map and then inserts them back in each others previous ID. I'm not sure if we even documents to retain their ids between operations, anywhere we use document ids we get an id and then immediately throw it away — we might be able to get away with just a `Vec`, as I can't imagine anyone ever needing even 20 documents open

**Comment by emilyyyylime** [comment]
> I'm working on implementing this, do you reckon having just a "move currently active document left/right" command would be good? And what keybinds should it use? I was thinking we could change "Select symbol references" to something else and use `<space>h` and `<space>l`

also if we do want a move left/right we could use the `[`/`]` menus. Maybe `]]` and `[[`?

**Comment by jabbalaci** [comment]
I would also need this feature. Moving a tab left / right would be enough for me. I don't know about the default keybinding but if there are two associated commands, I could map them to what I like. Moving a tab to the first and last place would be a plus, but moving to left / right would be enough for me.

**Comment by 7ombie** [comment]
`+ 1`

**Comment by mkellyxp** [comment]
I saw that this PR was closed and not merged.  Bit of a bummer, as it would def help me using the bufferline too.  I agree that it is more "GUI type editing" and using the buffer picker is probably better most of the time.  I know some people like myself are more visual and helps to have buffers opened in a certain order and I have:

C-right = [":bn"]
C-left = [":bp"]

Which feels amazing.. so I could imaging having C-shift-right = (move buffer right), etc.

At the end of the day, I trust the Helix team to make the right call, but would love to upvote this feature if possible.

**Comment by masonj5n** [comment]
I opened the PR originally but went out of town for a bit so didn't have a chance to give some thoughts. I agree that it would be a super helpful usability improvement for how I use Helix, I just find it a lot easier to manage and know which buffers I currently have open when I can see them at all times without opening the buffer picker and trying to read the full filenames there.

For me it's just a statusline where the info is open buffers, I don't even think they are clickable right? I definitely agree doing things to cater to a mouse-driven use-case shouldn't be the aim but I maybe disagree that this encourages that.

Love Helix and appreciate the team! This won't be a dealbreaker by any means but I do think for this workflow it would be awesome to have this feature, right now I sometimes close and reopen buffers to get them reordered haha.

**Comment by FrankensteinPillow** [comment]
I need this feature too +1

**Comment by 7ombie** [comment]
I often close and reopen buffers to reorder them as well. I also try to only open the files I need, eagerly closing any that I don't need open anymore, so I can navigate quickly between the open ones.

I exclusively rely on <kbd>Control</kbd>+<kbd>Left</kbd> and <kbd>Control</kbd>+<kbd>Right</kbd> to cycle through the 'tabs'. Being able to more easily group related files would really improve the experience.

**Comment by emilyyyylime** [comment]
#10077 was already closed by the team as "not fitting in" but I only saw it after implementing #11790 so decided to give it another chance anyways. Feel free to share your thoughts there

**Reply by edrozenberg** [reply]
I'd love this as well, too bad it sounds like the dev direction is not going that way.

I hope this at least doesn't happen:
https://github.com/helix-editor/helix/pull/10077#issuecomment-2043978343
> If anyrhing we should sort the bufferline the same as the buffer picker (altough there are some performance blockers for that right now).

Tab order that keeps changing during use. Nooooooooo ... :)
